src/Features/Core/Portable/PullMemberUp/AbstractCodeAndSolutionGenerator.cs

```diff
@@ -0,0 +1,336 @@
+            Document contextDocument,
+            CancellationToken cancellationToken = default)
+        {
+            CodeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var solution = contextDocument.Project.Solution;
+            var targetNodeSyntax = await CodeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(solution, result.Target).ConfigureAwait(false);
+
+            if (targetNodeSyntax != null)
+            {
+                var solutionEditor = new SolutionEditor(solution);
+                await InitializeEditorsAndSyntaxes(result, targetNodeSyntax, solutionEditor, solution, cancellationToken).ConfigureAwait(false);
+                if (result.Target.TypeKind == TypeKind.Interface)
+                {
+                    return GenerateTargetingInterfaceSolution(result, targetNodeSyntax, solutionEditor, solution);
+                }
+                else if (result.Target.TypeKind == TypeKind.Class)
+                {
+                    return GenerateTargetingClassSolution(result, targetNodeSyntax, solutionEditor, solution);
+                }
+                else
+                {
+                    throw new ArgumentException($"{nameof(result.Target)} should be interface or class");
+                }
+            }
+            else
+            {
+                return default;
+            }
+        }
+
+        public CodeAction GetCodeActionAsync(
+            AnalysisResult result,
+            Document contextDocument,
+            string title)
+        {
+            if (!result.IsValid)
+            {
+                return default;
+            }
+
+            CodeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            if (result.Target.TypeKind == TypeKind.Interface)
+            {
+                return CreateAddMembersToInterfaceAction(result, contextDocument, title);
+            }
+            else if (result.Target.TypeKind == TypeKind.Class)
+            {
+                return new SolutionChangeAction(
+                    title,
+                    async _ =>
+                    {
+                        var changedSolution = await GetSolutionAsync(result, contextDocument).ConfigureAwait(false);
+                        return changedSolution;
+                    });
+            }
+            else
+            {
+                throw new ArgumentException($"{nameof(result.Target)} should be interface or class");
+            }
+        }
+
+        private CodeAction CreateAddMembersToInterfaceAction(
+            AnalysisResult result,
+            Document contextDocument,
+            string title)
+        {
+            var options = new CodeGenerationOptions(generateMethodBodies: false, generateMembers: false);
+            var symbolsToPullUp = result.MembersAnalysisResults.
+            Select(analysisResult =>
```

Comment:
indentation is def wonky here.  this makes it look as if we're starting a statement with `Select(`